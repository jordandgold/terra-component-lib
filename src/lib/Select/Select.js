import React, { Component } from "react";
import "./Select.scss";
import PropTypes from "prop-types";
import ReactDOM from "react-dom";

class Select extends Component {
  constructor() {
    super();

    this.state = {
      deployed: false
    };

    this.selectOptionRefs = [];
  }

  toggleDeploy = () => {
    this.setState({
      deployed: !this.state.deployed
    });
  };

  handleKeyup = e => {
    if (!this.state.deployed) {
      return;
    }

    const node = this.selectOptionRefs.find(option => {
      return (
        option.current.props.option[0].toLowerCase() === e.key.toLowerCase()
      );
    });

    if (!node) {
      return;
    }

    const nodeLocation = ReactDOM.findDOMNode(node.current);

    window.scrollTo(0, nodeLocation.offsetTop);
  };

  createOptionRef = () => {
    const optionRef = React.createRef();

    this.selectOptionRefs.push(optionRef);

    return optionRef;
  };

  handleSelection = selection => {
    this.toggleDeploy();

    this.props.handleSelection(selection, this.props.name);
  };

  generateOptions = () => {
    return this.props.options.map((option, index) => {
      return (
        <SelectOption
          key={`${option}-${index}`}
          onClick={this.handleSelection}
          ref={this.createOptionRef()}
          option={option}
        />
      );
    });
  };

  render() {
    return (
      <div onKeyUp={e => this.handleKeyup(e)} tabIndex="0">
        <p className="ter-select__label" onClick={this.toggleDeploy}>
          {this.props.selection || this.props.defaultText}
        </p>
        <ul className={`select-options ${!this.state.deployed && "is-open"}`}>
          {this.generateOptions()}
        </ul>
      </div>
    );
  }
}

export class SelectOption extends Component {
  render() {
    const { option, onClick } = this.props;
    return (
      <li className="select-options__item" onClick={() => onClick(option)}>
        {option}
      </li>
    );
  }
}

export default Select;

Select.propTypes = {
  options: PropTypes.arrayOf(PropTypes.string).isRequired,
  handleSelection: PropTypes.func.isRequired,
  selection: PropTypes.string,
  defaultText: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired
};
