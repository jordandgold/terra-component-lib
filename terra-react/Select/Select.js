import _classCallCheck from "@babel/runtime/helpers/esm/classCallCheck";
import _createClass from "@babel/runtime/helpers/esm/createClass";
import _possibleConstructorReturn from "@babel/runtime/helpers/esm/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/esm/getPrototypeOf";
import _inherits from "@babel/runtime/helpers/esm/inherits";
import React, { Component } from "react";
import "./Select.scss";
import ReactDOM from "react-dom";
import Icon from "../Icon/Icon";

var Select =
/*#__PURE__*/
function (_Component) {
  _inherits(Select, _Component);

  function Select() {
    var _this;

    _classCallCheck(this, Select);

    _this = _possibleConstructorReturn(this, _getPrototypeOf(Select).call(this));

    _this.toggleClose = function (e) {
      if (_this.node.contains(e.target) || !_this.state.deployed) {
        return;
      }

      _this.toggleDeploy();
    };

    _this.toggleDeploy = function () {
      _this.setState({
        deployed: !_this.state.deployed
      });
    };

    _this.handleKeyup = function (e) {
      if (!_this.state.deployed) {
        return;
      }

      var nodes = Array.from(document.querySelectorAll(".ter-select__options-list-item"));
      var node = nodes.find(function (node) {
        return node.innerText[0].toLowerCase() === e.key.toLowerCase();
      });
      node.scrollIntoView();
    };

    _this.handleSelection = function (selection) {
      _this.toggleDeploy();

      _this.props.handleSelection(selection, _this.props.name);
    };

    _this.generateOptions = function () {
      return _this.props.options.map(function (option, index) {
        return React.createElement(SelectOption, {
          key: "".concat(option, "-").concat(index),
          onClick: _this.handleSelection,
          option: option
        });
      });
    };

    _this.state = {
      deployed: false
    };
    return _this;
  }

  _createClass(Select, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      document.addEventListener("mousedown", this.toggleClose);
    }
  }, {
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      document.removeEventListener("mousedown", this.toggleClose);
    }
  }, {
    key: "render",
    value: function render() {
      var _this2 = this;

      return React.createElement("div", {
        ref: function ref(node) {
          return _this2.node = node;
        },
        className: "ter-select ".concat(this.state.deployed && "is-open"),
        onKeyUp: function onKeyUp(e) {
          return _this2.handleKeyup(e);
        },
        tabIndex: "0"
      }, React.createElement("div", {
        className: "ter-select__selected",
        onClick: this.toggleDeploy
      }, this.props.selection || this.props.defaultText, React.createElement(Icon, {
        type: "open-caret-down-dark-16px",
        className: "ter-search-select__caret",
        size: "16px"
      })), React.createElement("ul", {
        className: "ter-select__options-list ".concat(this.state.deployed && "is-open")
      }, this.generateOptions()));
    }
  }]);

  return Select;
}(Component);

export var SelectOption =
/*#__PURE__*/
function (_Component2) {
  _inherits(SelectOption, _Component2);

  function SelectOption() {
    _classCallCheck(this, SelectOption);

    return _possibleConstructorReturn(this, _getPrototypeOf(SelectOption).apply(this, arguments));
  }

  _createClass(SelectOption, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          option = _this$props.option,
          _onClick = _this$props.onClick;
      return React.createElement("li", {
        className: "ter-select__options-list-item",
        onClick: function onClick() {
          return _onClick(option);
        }
      }, option);
    }
  }]);

  return SelectOption;
}(Component);
export default Select;