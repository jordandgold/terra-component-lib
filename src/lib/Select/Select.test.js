import React from "react";
import { shallow, mount } from "enzyme";
import Select, { SelectOption } from "./Select";
import ReactDOM from "react-dom";

describe("Select", () => {
  let wrapper;
  let mockHandleSelection;
  const mockOptions = ["cats", "dogs", "birds"];

  beforeEach(() => {
    mockHandleSelection = jest.fn();
    wrapper = shallow(
      <Select
        handleSelection={mockHandleSelection}
        options={mockOptions}
        selection={undefined}
        defaultText="Test default text"
        name="test name"
      />
    );
  });

  it("should match the snapshop", () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe("toggleDeploy", () => {
    it("should call setState", () => {
      wrapper.instance().toggleDeploy();

      expect(wrapper.state().deployed).toEqual(true);
    });

    it("should call toggleDeploy on click", () => {
      mockHandleSelection = jest.fn();
      wrapper = shallow(
        <Select
          handleSelection={mockHandleSelection}
          options={["dogs"]}
          selection={undefined}
          defaultText="Test default text"
          name="test name"
        />
      );
      const spy = jest.spyOn(wrapper.instance(), "toggleDeploy");
      wrapper.instance().forceUpdate();

      wrapper.find(".ter-select__selected").simulate("click");

      expect(spy).toHaveBeenCalled();
    });
  });

  describe("handleKeyup", () => {
    beforeEach(() => {
      mockHandleSelection = jest.fn();
      const alphaOptions = ["apple", "banana", "cat", "dog", "elephant"];
      const mockOptionRefs = [
        {
          current: {
            props: {
              option: "apple"
            }
          }
        },
        {
          current: {
            props: {
              option: "banana"
            }
          }
        },
        {
          current: {
            props: {
              option: "cat"
            }
          }
        },
        {
          current: {
            props: {
              option: "dog"
            }
          }
        },
        {
          current: {
            props: {
              option: "elephant"
            }
          }
        }
      ];

      wrapper = mount(
        <Select
          handleSelection={mockHandleSelection}
          options={alphaOptions}
          selection={undefined}
          defaultText="Test default text"
          name="test name"
        />
      );

      wrapper.instance().selectOptionRefs = mockOptionRefs;
    });

    const mockEventObject = { key: "c" };

    it("should call testKeyup", () => {
      const spy = jest.spyOn(wrapper.instance(), "handleKeyup");
      wrapper.instance().forceUpdate();

      wrapper.find(".ter-select").simulate("keyup", mockEventObject);

      expect(spy).toHaveBeenCalled();
    });

    it("should return if dropdown is not open", () => {
      const spy = jest.spyOn(wrapper.instance(), "handleKeyup");
      wrapper.instance().forceUpdate();

      wrapper.instance().handleKeyup(mockEventObject);

      expect(spy).toReturn();
    });

    it("should return if no matching nodes", () => {
      const noMatchEventObject = {
        key: "l"
      };
      const spy = jest.spyOn(wrapper.instance(), "handleKeyup");
      wrapper.instance().forceUpdate();

      wrapper.state().deployed = true;

      wrapper.instance().handleKeyup(noMatchEventObject);

      expect(spy).toReturn();
    });

    it("should call findDOMNode with the correct params", () => {
      const mockScrollTo = jest.fn();
      global.scrollTo = mockScrollTo;

      const mockOffsetTop = jest.fn();
      ReactDOM.findDOMNode = jest.fn().mockImplementation(() => {
        return {
          offsetTop: mockOffsetTop
        };
      });

      wrapper.state().deployed = true;

      wrapper.instance().handleKeyup(mockEventObject);

      expect(ReactDOM.findDOMNode).toHaveBeenCalled();
    });

    it("should call window.scrollTo", () => {
      const mockScrollTo = jest.fn();
      global.scrollTo = mockScrollTo;

      const mockOffsetTop = jest.fn();
      ReactDOM.findDOMNode = jest.fn().mockImplementation(() => {
        return {
          offsetTop: mockOffsetTop
        };
      });

      wrapper.state().deployed = true;

      wrapper.instance().handleKeyup(mockEventObject);

      expect(mockScrollTo).toHaveBeenCalled();
    });
  });

  describe("createOptionRef", () => {
    it("should call createRef", () => {
      const mockCreateRef = jest.fn();

      React.createRef = mockCreateRef;

      wrapper.instance().createOptionRef();

      expect(mockCreateRef).toHaveBeenCalled();
    });

    it("should should push a new ref", () => {
      const mockCreateRef = jest.fn().mockImplementation(() => {
        return "test";
      });
      React.createRef = mockCreateRef;
      wrapper.instance().selectOptionRefs = [];
      const expected = ["test"];

      wrapper.instance().createOptionRef();

      expect(wrapper.instance().selectOptionRefs).toEqual(expected);
    });

    it("should return an optionRef", () => {
      const mockCreateRef = jest.fn().mockImplementation(() => {
        return "test";
      });
      React.createRef = mockCreateRef;
      wrapper.instance().selectOptionRefs = [];
      const expected = "test";

      const response = wrapper.instance().createOptionRef();

      expect(response).toEqual(expected);
    });
  });

  describe("handleSelection", () => {
    it("should call toggleDeploy", () => {
      const spy = jest.spyOn(wrapper.instance(), "toggleDeploy");
      wrapper.instance().forceUpdate();

      wrapper.instance().handleSelection("cake");

      expect(spy).toHaveBeenCalled();
    });

    it("should call handleSelection from props", () => {
      wrapper.instance().handleSelection("yum");

      expect(mockHandleSelection).toHaveBeenCalledWith("yum", "test name");
    });
  });

  describe("generateOptions", () => {
    it("should return some JSX", () => {
      const expected = [
        '<li class="ter-select__options-list-item">cats</li>',
        '<li class="ter-select__options-list-item">dogs</li>',
        '<li class="ter-select__options-list-item">birds</li>'
      ];
      const result = wrapper
        .instance()
        .generateOptions()
        .map(option => {
          return shallow(option).html();
        });

      expect(result).toEqual(expected);
    });
  });

  describe("SelectOption", () => {
    let selectOptionWrapper;
    let mockOnClick;

    beforeEach(() => {
      mockOnClick = jest.fn();
      selectOptionWrapper = shallow(
        <SelectOption option="kitty" onClick={mockOnClick} />
      );
    });

    it("should match the snapshot", () => {
      expect(selectOptionWrapper).toMatchSnapshot();
    });

    it("should call onClick when clicked", () => {
      selectOptionWrapper.find("li").simulate("click");

      expect(mockOnClick).toHaveBeenCalledWith("kitty");
    });
  });
});
