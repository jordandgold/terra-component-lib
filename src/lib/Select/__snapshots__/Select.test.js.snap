// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select SelectOption should match the snapshot 1`] = `
<li
  className="ter-select__options-list-item"
  onClick={[Function]}
>
  kitty
</li>
`;

exports[`Select should match the snapshop 1`] = `
<div
  className="ter-select is-open"
  onKeyUp={[Function]}
  tabIndex="0"
>
  <div
    className="ter-select__selected"
    onClick={[Function]}
  >
    Test default text
    <Icon
      className="ter-search-select__caret"
      size="16px"
      type="open-caret-down-dark-16px"
    />
  </div>
  <ul
    className="ter-select__options-list is-open"
  >
    <SelectOption
      key="cats-0"
      onClick={[Function]}
      option="cats"
    />
    <SelectOption
      key="dogs-1"
      onClick={[Function]}
      option="dogs"
    />
    <SelectOption
      key="birds-2"
      onClick={[Function]}
      option="birds"
    />
  </ul>
</div>
`;
